libraries/AP_Soaring/SoaringController.cpp

```diff
@@ -0,0 +1,407 @@
+    
+    _ahrs.get_position(_prev_update_location);
+    _prev_update_time = AP_HAL::micros64();
+    _thermal_start_time_us = AP_HAL::micros64();
+}
+   
+void SoaringController::init_cruising()
+{
+    _cruise_start_time_us = AP_HAL::micros64();
+    // Start glide. Will be updated on the next loop.
+    _throttle_suppressed = true;
+}
+
+void SoaringController::update_thermalling(float loiter_radius)
+{
+    _loiter_rad = loiter_radius;
+    struct Location current_loc;
+    _ahrs.get_position(current_loc);
+
+    if (_new_data) {      
+        Vector2f diff = location_diff(_prev_update_location, current_loc); // get distances from previous update
+        float dx = diff.x;
+        float dy = diff.y;
+
+        // Wind correction
+        Vector3f wind = _ahrs.wind_estimate();
+        float dx_w = wind.x * (AP_HAL::micros64() - _prev_update_time) * 1e-6;
+        float dy_w = wind.y * (AP_HAL::micros64() - _prev_update_time) * 1e-6;
+        dx -= dx_w;
+        dy -= dy_w;
+
+        if (0) {
+            // Print filter info for debugging
+            int i;
+            for (i = 0; i < 4; i++) {
+                hal.console->printf("%e ", (double)_ekf.P[i][i]);
+            }
+            for (i = 0; i< 4; i++) {
+                hal.console->printf("%e ", (double)_ekf.X[i]);
+            }
+        } else {
+             // write log - save the data.
+            log_tuning.time_us = AP_HAL::micros64();
+            log_tuning.netto_rate = _vario_reading;
+            log_tuning.dx = dx;
+            log_tuning.dy = dy;
+            log_tuning.x0 = _ekf.X[0];
+            log_tuning.x1 = _ekf.X[1];
+            log_tuning.x2 = _ekf.X[2];
+            log_tuning.x3 = _ekf.X[3];
+            log_tuning.lat = current_loc.lat;
+            log_tuning.lng = current_loc.lng;
+            log_tuning.alt = _alt;
+            log_tuning.dx_w = dx_w;
+            log_tuning.dy_w = dy_w;
+            log_tuning.nsamples = _nsamples;
+        }
+        log_data(); 
+        _ekf.update(_vario_reading,dx, dy);       // update the filter
+         
+        _prev_update_location = current_loc;      // save for next time
+        _prev_update_time = AP_HAL::micros64();
+        _new_data = false;
+    }
+}
+
+void SoaringController::update_cruising()
+{
+    if (is_active() && suppress_throttle()) {
+        init_cruising();
+    }
+}
+
+void SoaringController::update_vario()
+{
+    Location current_loc;
+    _ahrs.get_position(current_loc);
+    _alt = (current_loc.alt -_ahrs.get_home().alt) / 100.0f;
+    
+    if (fabsf(_alt - _last_alt) > 0.0001f) { // if no change in altitude then there will be no update of ekf buffer
+        // Both filtered total energy rates and unfiltered are computed for the thermal switching logic and the EKF
+        float aspd;
+        float roll = _ahrs.roll;
+        if (!_ahrs.airspeed_estimate(&aspd)) {
+            aspd = 0.5f * (_aparm.airspeed_min + _aparm.airspeed_max); //
```

Comment:
+1 to using the trim as the better value for the unconfigured case. It's the highest chance of being set reasonable. I've seen a number of logs where MIN/MAX was never moved from the default values which could lead to unexpected behavior.